xdas/compose.py

```diff
@@ -0,0 +1,530 @@
+from collections import Counter, UserDict
+from collections.abc import Callable, Hashable
+from copy import copy
+from typing import Any, Self, Type
+
+from .database import Database
+from .processing import DatabaseLoader, DatabaseWriter, ProcessingChain
+
+ChainType = Type[ProcessingChain]
+LoaderType = Type[DatabaseLoader]
+WriterType = Type[DatabaseWriter]
+DatabaseType = Type[Database]
+
+
+class Sequence(UserDict):
+    """
+    A class to handle a sequence of operations.
+    Each operation is represented by an Atom class object,
+    which contains the function and its arguments.
+    For stateful operations, use the StateAtom class.
+
+    Sequence inherits from UserDict, and therefore behaves
+    as a dictionary. It is extended with convenience
+    methods to reorder the dictionary. (By default, all
+    dict objects in Python >= 3.7 maintain their insertion order).
+
+    Attributes
+    ----------
+    name_counter : Counter
+        Counter object to keep track of (duplicate)
+        dictionary keys
+
+    Examples
+    --------
+    >>> op1 = xdas.Atom(xdas.signal.taper, dim="time")
+    ... op2 = xdas.Atom(xdas.signal.taper, dim="space", name="spatial taper")
+    ... op3 = xdas.Atom(numpy.square)
+    ... op4 = xdas.Atom(my_func, arg1=1.0, arg2=[0, 1, 2])
+    ... sequence = xdas.Sequence(op1, op2, op3, op4)
+    ... sequence["abs"] = xdas.Atom(np.abs)
+    ... del sequence["spatial taper"]
+    ... print(sequence)  # Print a summary of the Sequence
+    ... sequence.execute(db)
+
+    """
+
+    def __init__(self, *args: Any) -> None:
```

Comment:
This would require the user to provide `args` in the form of an iterable, right? So `sequence = xdas.Sequence(op1, op2, op3)` would become `sequence = xdas.Sequence([op1, op2, op3])`?

Reply:
Indeed.

Reply:
So we stick to `*args`, which in my opinion looks nicer?